dockstore-webservice/src/main/java/io/openapi/api/impl/ToolsApiServiceImpl.java

```diff
@@ -360,45 +359,45 @@ public Response toolsGet(String id, String alias, String toolClass, String descr
-            // construct links to other pages
-            List<String> filters = new ArrayList<>();
-            handleParameter(id, "id", filters);
-            handleParameter(organization, "organization", filters);
-            handleParameter(name, "name", filters);
-            handleParameter(toolname, "toolname", filters);
-            handleParameter(description, "description", filters);
-            handleParameter(author, "author", filters);
-            handleParameter(registry, "registry", filters);
-            handleParameter(String.valueOf(actualLimit), "limit", filters);
-
-            final long numPages = (numEntries.sum()) / actualLimit;
-
-            if (startIndex + actualLimit < numEntries.sum()) {
-                URI nextPageURI = new URI(config.getExternalConfig().getScheme(), null, config.getExternalConfig().getHostname(), port,
-                    ObjectUtils.firstNonNull(config.getExternalConfig().getBasePath(), "") + relativePath,
-                    Joiner.on('&').join(filters) + "&offset=" + (offsetInteger + 1), null).normalize();
-                responseBuilder.header("next_page", nextPageURI.toURL().toString());
-            }
-            URI lastPageURI = new URI(config.getExternalConfig().getScheme(), null, config.getExternalConfig().getHostname(), port,
-                ObjectUtils.firstNonNull(config.getExternalConfig().getBasePath(), "") + relativePath, Joiner.on('&').join(filters) + "&offset=" + numPages, null)
-                .normalize();
-            responseBuilder.header("last_page", lastPageURI.toURL().toString());
+        responseBuilder.header("self_link", createUrlString(scheme, hostname, port, path, query));
+        if (startIndex + actualLimit < numEntries.sum()) {
+            responseBuilder.header("next_page", createUrlString(scheme, hostname, port, path, positionQuery(query, actualLimit, offsetInteger + 1L)));
+        }
+        final long numPages = numEntries.sum() / actualLimit;
+        responseBuilder.header("last_page", createUrlString(scheme, hostname, port, path, positionQuery(query, actualLimit, numPages)));
+
+        return responseBuilder.build();
+    }
 
+    private String createUrlString(String scheme, String hostname, int port, String path, String query) {
+        try {
+            return new URI(scheme, null, hostname, port, path, query, null).normalize().toURL().toString();
         } catch (URISyntaxException | MalformedURLException e) {
-            throw new CustomWebApplicationException("Could not construct page links", HttpStatus.SC_BAD_REQUEST);
+            throw new CustomWebApplicationException("Could not create url string", HttpStatus.SC_BAD_REQUEST);
         }
-        trsListener.loadTRSResponse(hashcode, responseBuilder);
-        return responseBuilder.build();
+    }
+
+    private String positionQuery(String query, long limit, long offset) {
```

Comment:
The httpclient library (on our classpath already) has some library methods that can simplify + document this. 
For example, https://stackoverflow.com/questions/2809877/how-to-convert-map-to-url-query-string/37744000#37744000

Reply:
When I was authoring this PR, I briefly tried the httpclient library solution, but the resulting implementation was not shorter or more simple, and it subjected our query string to a bunch of unnecessary encoding/decoding steps along the way, which gave me pause.  What we really want is something super-simple that adds/replaces the two field while passing through the rest of the query string exactly as it was.  Maybe one of the value fields includes some characters in an unusual encoding that our browser/server understands, but httpclient does not... 

How's about I add a couple of comments to clarify what is happening, and include a pointer to the httpclient library (for anyone who might need to implement something more complicated in the future)?